lib/evmone/instructions.hpp

```diff
@@ -375,16 +487,18 @@ inline void address(StackTop stack, ExecutionState& state) noexcept
 
 inline Result balance(StackTop stack, int64_t gas_left, ExecutionState& state) noexcept
 {
-    auto& x = stack.top();
-    const auto addr = intx::be::trunc<evmc::address>(x);
+    auto& x = stack[0];
+    const auto addr = intx::be::trunc<evmc::address>(x.val);
 
     if (state.rev >= EVMC_BERLIN && state.host.access_account(addr) == EVMC_ACCESS_COLD)
     {
         if ((gas_left -= instr::additional_cold_account_access_cost) < 0)
             return {EVMC_OUT_OF_GAS, gas_left};
     }
 
-    x = intx::be::load<uint256>(state.host.get_balance(addr));
+    x.val = intx::be::load<uint256>(state.host.get_balance(addr));
+    // TODO do we need some constraint on gas here? probably not...
```

Comment:
I do not think that we will need any requirements on gas, but we are missing an "exact" match requirement on `addr` (if revision >= EVMC_BERLIN), because gas calculation depends on it. Technically we do not need "exact" match, only that `state.host.access_account(addr) == EVMC_ACCRESS_COLD` matches. But I think exact match is fine.